File: apis/src/test/java/com/beat/apis/ticket/facade/TicketFacadeTest.java
package com.beat.apis.ticket.facade;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.beat.apis.schedule.api.type.ScheduleNumberType;
import com.beat.apis.ticket.application.command.TicketCommandService;
import com.beat.apis.ticket.application.query.TicketListQuery;
import com.beat.apis.ticket.application.query.TicketQueryService;
import com.beat.apis.ticket.application.result.TicketRetrieveResult;
import com.beat.apis.booking.api.type.BookingStatusType;

@ExtendWith(MockitoExtension.class)
class TicketFacadeTest {

	@Mock
	private TicketQueryService ticketQueryService;

	@Mock
	private TicketCommandService ticketCommandService;

	private TicketFacade ticketFacade;

	@BeforeEach
	void setUp() {
		ticketFacade = new TicketFacade(ticketQueryService, ticketCommandService);
	}

	@Test
	void findTicketsDelegatesToService() {
		Long memberId = 1L;
		Long performanceId = 100L;
		List<ScheduleNumberType> scheduleNumbers = List.of(ScheduleNumberType.FIRST);
		List<BookingStatusType> bookingStatuses = List.of(BookingStatusType.CHECKING_PAYMENT);
		TicketListQuery query = new TicketListQuery(null, List.of("FIRST"), List.of("CHECKING_PAYMENT"));
		TicketRetrieveResult expected = new TicketRetrieveResult("title", "team", 1, 100, 10, List.of());
		when(ticketQueryService.findAllTicketsByConditions(memberId, performanceId, query))
			.thenReturn(expected);

		ticketFacade.findTickets(memberId, performanceId, scheduleNumbers, bookingStatuses);

		verify(ticketQueryService).findAllTicketsByConditions(memberId, performanceId, query);
	}

	@Test
	void searchTicketsDelegatesToService() {
		Long memberId = 1L;
		Long performanceId = 100L;
		String searchWord = "홍길동";
		List<ScheduleNumberType> scheduleNumbers = List.of(ScheduleNumberType.FIRST);
		List<BookingStatusType> bookingStatuses = List.of(BookingStatusType.CHECKING_PAYMENT);
		TicketListQuery query = new TicketListQuery(searchWord, List.of("FIRST"), List.of("CHECKING_PAYMENT"));
		TicketRetrieveResult expected = new TicketRetrieveResult("title", "team", 1, 100, 10, List.of());
		when(ticketQueryService.searchAllTicketsByConditions(memberId, performanceId, query)).thenReturn(expected);

		ticketFacade.searchTickets(memberId, performanceId, searchWord, scheduleNumbers, bookingStatuses);

		verify(ticketQueryService).searchAllTicketsByConditions(memberId, performanceId, query);
	}
}

File: apis/src/test/kotlin/com/beat/apis/ticket/facade/TicketFacadeTest.kt
package com.beat.apis.ticket.facade

import com.beat.apis.booking.api.type.BookingStatusType
import com.beat.apis.schedule.api.type.ScheduleNumberType
import com.beat.application.frontoffice.ticket.maker.query.TicketListQuery
import com.beat.application.frontoffice.ticket.maker.query.TicketQueryService
import com.beat.application.frontoffice.ticket.maker.query.TicketRetrieveResult
import com.beat.application.frontoffice.ticket.maker.command.TicketCommandService
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class TicketFacadeTest {

    private val ticketQueryService = Mockito.mock(TicketQueryService::class.java)
    private val ticketCommandService = Mockito.mock(TicketCommandService::class.java)
    private val ticketFacade = TicketFacade(ticketQueryService, ticketCommandService)

    @Test
    fun `find tickets delegates with api filters mapped to application query`() {
        val memberId = 1L
        val performanceId = 100L
        val scheduleNumbers = listOf(ScheduleNumberType.FIRST)
        val bookingStatuses = listOf(BookingStatusType.CHECKING_PAYMENT)
        val query = TicketListQuery(null, listOf("FIRST"), listOf("CHECKING_PAYMENT"))
        val expected = TicketRetrieveResult("title", "team", 1, 100, 10, emptyList())
        Mockito.`when`(ticketQueryService.findAllTicketsByConditions(memberId, performanceId, query))
            .thenReturn(expected)

        ticketFacade.findTickets(memberId, performanceId, scheduleNumbers, bookingStatuses)

        Mockito.verify(ticketQueryService).findAllTicketsByConditions(memberId, performanceId, query)
    }

    @Test
    fun `search tickets delegates with api filters mapped to application query`() {
        val memberId = 1L
        val performanceId = 100L
        val searchWord = "홍길동"
        val scheduleNumbers = listOf(ScheduleNumberType.FIRST)
        val bookingStatuses = listOf(BookingStatusType.CHECKING_PAYMENT)
        val query = TicketListQuery(searchWord, listOf("FIRST"), listOf("CHECKING_PAYMENT"))
        val expected = TicketRetrieveResult("title", "team", 1, 100, 10, emptyList())
        Mockito.`when`(ticketQueryService.searchAllTicketsByConditions(memberId, performanceId, query))
            .thenReturn(expected)

        ticketFacade.searchTickets(memberId, performanceId, searchWord, scheduleNumbers, bookingStatuses)

        Mockito.verify(ticketQueryService).searchAllTicketsByConditions(memberId, performanceId, query)
    }
}
